File: stack_queues/496_next_greater_element_i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:

        stack = []
        result = [-1] * len(nums1)

        for index, val in enumerate(nums1):
            for j in nums2:
                if j == val:
                    stack.append(j)
                else:
                    if stack and j > stack[-1]:
                        result[index] = j
                        break
            stack.clear()
                    
        return result
        

File: monotonic_stack/496_next_greater_element_i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        int[] res = new int[nums1.length];
        Stack<Integer> stack = new Stack<>();
        HashMap<Integer, Integer> map = new HashMap<>();

        for (int i = 0; i < nums2.length; i++) {
            while (!stack.empty() && nums2[stack.peek()] < nums2[i]) {
                int topIndex = stack.pop();
                map.put(nums2[topIndex] , nums2[i]);
            }
            stack.push(i);
        }

        while (!stack.empty()) {
            map.put(nums2[stack.pop()], -1);
        }
       
        for (int i = 0; i < nums1.length; i++) {
            res[i] = map.get(nums1[i]);
        }
        return res;
    }
}